llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +135,170 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
+  }
+  return Hash;
+}
+
+static std::pair<int, short> trackIntrinsicChain(const Value *V) {
+  int CountIntrinsicCall = 0;
+  hash_code Hash(0);
+  if (!V->getType()->isPointerTy())
+    return {-1, 0};
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+
+    if (auto *GEP = dyn_cast<GEPOperator>(V)) {
+      V = GEP->getPointerOperand();
+    } else if (Operator::getOpcode(V) == Instruction::BitCast ||
+               Operator::getOpcode(V) == Instruction::AddrSpaceCast) {
+      V = cast<Operator>(V)->getOperand(0);
+      if (!V->getType()->isPointerTy())
+        break;
+    } else {
+      if (auto *PHI = dyn_cast<PHINode>(V)) {
+        // Reached final destination;
+        return {CountIntrinsicCall, Hash};
+      } else if (auto *AIEObject = lookThroughAIEAddressingModes(V)) {
+        hash_code NewHash = hashIntrinsic(V);
+        // Initialize or compare hash?
+        if (Hash == hash_code(0)) {
+          Hash = NewHash;
+        } else if (Hash != NewHash) {
+          // Iterator information mismatch between sequence of intrinsic
+          // calls. We stop here.
+          return {-1, 0};
+        }
+        V = AIEObject;
+        CountIntrinsicCall++;
+        continue;
+      } else {
+        break;
+      }
+    }
+    assert(V->getType()->isPointerTy() && "Unexpected operand type!");
+  }
+  return {-1, 0};
+}
+
+static AliasResult aliasAIEIntrinsic(const MemoryLocation &LocA,
+                                     const MemoryLocation &LocB,
+                                     const Value *BaseA, const Value *BaseB) {
+
+  const PHINode *PhiA = dyn_cast<PHINode>(BaseA);
+  const PHINode *PhiB = dyn_cast<PHINode>(BaseB);
+
+  // Both bases must be PHI nodes.
+  if (!PhiA || !PhiB)
+    return AliasResult::MayAlias;
+
+  // Skip non-pointer values (cast from pointers?).
+  if (!PhiA->getType()->isPointerTy() || !PhiB->getType()->isPointerTy())
+    return AliasResult::MayAlias;
+
+  // Our target is loop-recurrent PHI nodes.
+  if (PhiA->getNumIncomingValues() != 2 || PhiB->getNumIncomingValues() != 2)
+    return AliasResult::MayAlias;
+
+  // Both PHI nodes must be in the same MBB.
+  if (PhiA->getParent() != PhiB->getParent())
+    return AliasResult::MayAlias;
+
+  auto GetExternalIndex = [](const PHINode *Phi) {
+    return Phi->getIncomingBlock(0) == Phi->getParent() ? 1 : 0;
+  };
+
+  auto GetInternalIndex = [](const PHINode *Phi) {
+    return Phi->getIncomingBlock(0) == Phi->getParent() ? 0 : 1;
+  };
+
+  const Value *ExtPhiValA = PhiA->getIncomingValue(GetExternalIndex(PhiA));
+  const Value *ExtPhiValB = PhiB->getIncomingValue(GetExternalIndex(PhiB));
+
+  // Both Phis must have the same value as an external incoming operand,
+  // so we guarantee that we have two new pointers rooted in the same
+  // original pointer.
+  //
+  // NOTE: We don't care if A and B are the same or not, so we can also
+  // desambiguate different steps starting from from the same pointer (Phi).
```

Comment:
nit: *dis*ambiiguate

Reply:
Oh, this is always a confusion in my head because of my first language ;-)